### ISSUE TITLE
Use "withImmutable" for Immutable argument properties in callbacks

### ISSUE THREAD

**Issue by kettanaito** [kettanaito/react-advanced-form#95]
## What
Need to allow the usage of Immutable instances in callback argument properties.

## Why
* `ImmutableJS` is a required peer dependency. There is a high chance project is using Immutable already, or going to do so.
* Conversion to plain JavaScript instances takes resources and time. Unnecessary when Immutable workflow is inherited in the project.

## How
1. Conversion to mutable **still happens by default**. RAF is never enforcing to use any technology.
2. Introduce the prop controlling this behavior (on a `FormProvider` level):

```jsx
<FormProvider withImmutable />
```

3. Introduce a helper function to keep conditional logic at one place:

```js
function emitMutable(immutableInstance) {
  return withImmutable ? immutableInstance : immutableInstance.toJS();
}
```

4. Find a way for the helper function to access `this.context.withImmutable` automatically, without the need to drag it around to each util method.

**Comment by kettanaito** [comment]
Since RAF requires `immutable` as a peer dependency, there is a high change Immutable is already used in the project. Therefore, conversion `toJS()` becomes an unnecessary performance overkill.

**Comment by kettanaito** [comment]
I suggest to strongly consider this change because `immutable` is already a required peer dependency. There is a very high chance the project may use immutable instances for internal purposes, so it's not confusing how to work with them.

### Suggestion
```jsx
import { FormProvider } from 'react-advanced-form';

const renderApp = () => (
  <FormProvider withImmutable />
);
```

**Comment by kettanaito** [comment]
Prop naming suggestions:
- `withImmutable`
- `useImmutable`
- `mutableCallbacks` (expects `false` to omit conversion to mutable)

**Comment by kettanaito** [comment]
The implementation is merged just now.

Need to update the docs.

**Comment by kettanaito** [comment]
## Documentation
Updated [FormProvider](https://kettanaito.gitbooks.io/react-advanced-form/components/Provider.html) documentation to include `withImmutable `definition and usage examples.

**Comment by kettanaito** [comment]
Published under [`0.19.10`](https://github.com/kettanaito/react-advanced-form/commit/5a74ff724e70815fbe7e43c554c9ad391e9144e2). Closing.
